File: Java/Tree/InvertBinaryTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class InvertBinaryTree {
    public TreeNode invertTree(TreeNode root) {
        if(root==null) return null;
        TreeNode oldLeft = root.left;
        root.left = inertTree(root.right);
        root.right = invertTree(oldLeft);
        return root;
    }

    // helper
   	public TreeNode invertTree(TreeNode root){
   		if(root==null) return root;
   		invertTree(root);
   		return root;
   	}
    public void invertTreeHelper(TreeNode root){
    	if(root==null) return;
    	TreeNode tmp = root.left;
    	root.left = root.right;
    	root.right = tmp;
    	invertTree2(root.left);
    	invertTree2(root.right);
    }
}

File: Sum/Tree/invertbinarytree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
 // recursive solution, but since it is recursive, it is not scalable, the stack tend to overflow. 
var invertTree1 = function(root) {
  	if(root===null) return root;
    var tmpNode = root.left;
    root.left = root.right;
    root.right = tmpNode;
    invertTree(root.left);
    invertTree(root.right);
    return root;
};

// to optimize it to make it more scalable, we can change this to a iterative way, using a stack 

var invertTree2 = function(root) {
    if(root===null) return root;
    var stack = [];
    stack.push(root);
    while(stack.length>0){
        var curRt = stack.pop();
        var tmpLeft = curRt.left;
        curRt.left = curRt.right;
        curRt.right = tmpLeft;
        if(curRt.left!==null) stack.push(curRt.left);
        if(curRt.right!==null) stack.push(curRt.right);
    }
    return root;
};

/*
  and since when we push to stack, the children have already been swapped, so we can change it easily to BFS,
  i.e level order traversal 
*/

var invertTree3 = function(root) {
	if(root===null) return root;
	var queue = [];
	queue.unshift(root);
	while(queue.length>0){
		var curRt = queue.shift();
		var tmpLeft = curRt.left;
		curRt.left = curRt.right;
		curRt.right = tmpLeft;
		if(curRt.left!==null) queue.unshift(curRt.left);
		if(curRt.right!==null) queue.unshift(curRt.right);
	}
	return root;
};